[CMake] Improve SOFA installation and packaging

guparan opened the pull request:
This PR contains all the changes I did to produce SOFA releases.
Basically it is almost only CMake improvements to ease SOFA installation.

Some of them may not be clean/wanted/useful, **please give feedback**.

WARNING: I renamed some macros in SofaMacros.cmake so this could break your code.

______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

Commit:
[CMake] FIX SOFA install mishandling of glut dependency and libpng

The flag SOFA_HAVE_GLUT_GUI is replaced by SOFAGUI_GLUT
Only the SOFAGUI requires glut.
IF glut found does not mean to add it as a dependency.

Libpng is also now well shipped for Linux.

Commit:
[CMake] FIX install for MacOS

guparan committed:
[CPack] ADD packing rules

- COMPONENT param in INSTALL rules
- NSIS rules to generate a Windows installer

guparan committed:
[CMake] CLEAN install and copy rules for dependencies

guparan committed:
[CMake] CLEAN install rules for headers

guparan committed:
[CMake] FIX wrong file name for license

guparan committed:
[CMake] REMOVE CPack options from SofaKernel

guparan committed:
[CPack] FIX install directory

guparan committed:
[CMake] FIX shared libs install directory

guparan committed:
[CMake] FIX RPATH by using $ORIGIN

guparan committed:
[CMake] FIX symlinks copy for libs + RENAME macros

sofa_install_get_libraries -> sofa_install_files_from_library
sofa_copy_extlib -> sofa_copy_files_from_target
sofa_install_extlib -> sofa_install_files_from_target

guparan committed:
[CMake] ADD libicu* install for UNIX

guparan committed:
[CMake] FIX libs (symlinks) detection on install

guparan committed:
[CImgPlugin] EDIT CMake: remove RPATH (Unix)

guparan committed:
[CMake] FIX missing Boost libs on install

guparan committed:
[CMake] EDIT: no more Qt libs install for Unix

User will handle Qt himself by installing
qt5-default, qttools5-dev-tools, libqt5opengl5-dev

guparan committed:
[CMake] FIX loop on Boost_LIBRARIES

guparan committed:
[CMake] REMOVE Boost autodetect

This function leads to misunderstand which Boost is being used.
The only variable to set in CMake to locate Boost is BOOST_LIBRARYDIR

guparan committed:
[CPack] Improve Windows installer

guparan committed:
[CPack] FIX MacOS deployment

guparan committed:
[CMake] ADD Qt::DBus dependency for Windows deployment

guparan committed:
[CMake] FIX resources installation

guparan committed:
[CMake] REMOVE duplicated install rule

guparan committed:
Merge remote-tracking branch 'upstream/master' into cmake_cpack_rules

guparan committed:
Merge remote-tracking branch 'upstream/master' into cmake_cpack_rules

# Conflicts:
#	SofaKernel/SofaFramework/CMakeLists.txt

guparan committed:
[CMake] ADD gtest and Eigen to install

guparan committed:
[CMake] FIX Boost install and copy

guparan committed:
FIX header install + REMOVE useless PUBLIC_HEADER

guparan committed:
[CMake] REMOVE useless Qt instructions

guparan committed:
[CMake] RENAME macros

guparan committed:
[CPack] ADD GTest headers to installer

guparan committed:
[CMake] REMOVE gtest from SOFA_INCLUDE_EXTLIBS

was added in 8126f983a50aebca3ec96df04c4766d5bb425eec

guparan committed:
[CMake] FIX Boost copy and install

guparan committed:
[CMake] FIX Qt copy and install

Qt dependencies are not the same in SofaGuiQt and QGLViewer

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_cpack_rules

# Conflicts:
#	CMakeLists.txt
#	SofaKernel/SofaFramework/CMakeLists.txt
#	SofaKernel/framework/sofa/simulation/CMakeLists.txt

guparan committed:
[CMake] ADD component to install rule

guparan committed:
[CMake] FIX boost::program_options link

guparan committed:
[CImgPlugin] FIX PNG link

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_cpack_rules

guparan committed:
[CMake] ADD missing COMPONENT in install rules

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_cpack_rules

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_cpack_rules

# Conflicts:
#	applications/plugins/CImgPlugin/CMakeLists.txt
#	applications/projects/runSofa/CMakeLists.txt
#	applications/sofa/gui/qt/CMakeLists.txt

guparan committed:
[CMake] ADD missing COMPONENT in install rules

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_cpack_rules

guparan committed:
[SofaPython] EDIT: move resources install to SofaMacros.cmake

guparan committed:
[CMake] EDIT: remove default install of git_version

guparan committed:
[CImgPlugin] ADD libtiff and libjpeg install

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_cpack_rules

guparan committed:
[CMake] ADD custom.cmake variant for packaging

guparan committed:
[CMake] Force PSL OFF for packaging

guparan committed:
[runSofa] FIX CMakeLists RESOURCE_FILES

guparan committed:
[CImgPlugin] REMOVE RPATH in CMakeLists

guparan committed:
[SofaHelper] FIX Utils for MacOS bundle

- msg_error in computeSofaPathPrefix was triggering a segfault
- getSofaPathTo needs to check that the path exists

guparan committed:
[runSofa] EDIT: use PluginRepository for configPlugin file finding

guparan committed:
[runSofa] Remove redundant resources install

guparan committed:
[runSofa] ADD plugins in MacOS bundle + FIX bundling

Force dependencies to CImgPlugin, SofaMiscCollision and SofaPython to have them in the .app Bundle

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_cpack_rules

guparan committed:
Merge remote-tracking branch 'guparan/fix_pluginmanager_gui' into cmake_cpack_rules

guparan committed:
[CMake] FIX MacOS bundle

guparan committed:
[CMake] Enable SOFAALLCOMMONCOMPONENTS in package.cmake

guparan committed:
[SofaAllCommonComponents] ADD install rule

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_cpack_rules

guparan committed:
[package] Remove Windows specific code

guparan committed:
[CPack] FIX install dir on Windows

Default dir is now %USERPROFILE%

# Conflicts:
#	CMakeLists.txt

guparan committed:
[SofaMiscCollision] FIX SofaSphFluid link

SofaSphFluid was linked even when the plugin was disabled.

guparan committed:
[CMake] EDIT: disable unwanted plugins

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_cpack_rules

# Conflicts:
#	CMakeLists.txt
#	SofaKernel/SofaFramework/CMakeLists.txt
#	applications/plugins/CImgPlugin/CMakeLists.txt
#	applications/plugins/SofaMiscCollision/CMakeLists.txt
#	applications/projects/runSofa/CMakeLists.txt
#	applications/projects/runSofa/Main.cpp

guparan committed:
[CImgPlugin] CLEAN CMakeLists indentation

guparan committed:
[SofaFramework] FIX typo in CMakeLists

guparan committed:
[SofaFramework] FIX duplication in CMakeLists

guparan committed:
[runSofa] RENAME postInstall to bundle-post

guparan committed:
[runSofa] EDIT bundle-post message

guparan committed:
[SofaFramework] CLEAN macros order in SofaMacros

guparan committed:
[SofaFramework] CLEAN boost::chrono and boost::atomic deps

guparan committed:
[runSofa] FIX MacOS bundle

guparan committed:
[CMake] EDIT: enable GTest in package.cmake

guparan committed:
[CMake] EDIT: improve copy/install macros

Use ARGN to loop over arguments (no more need to put argument between quotes).

guparan commented:
Have a look at https://github.com/sofa-framework/sofa/pull/635/commits/6e0424065133651914561620a7952b34447c1ed9, I think it is the only commit affecting actual code (not CMake).

guparan commented:
[ci-build][with-scene-tests]

jnbrunet commented on SofaKernel/framework/sofa/helper/Utils.cpp:
```diff
@@ -249,7 +249,7 @@ static std::string computeSofaPathPrefix()
         std::size_t pos = exePath.rfind("/bin/");
         if (pos == std::string::npos) {
-            msg_error("Utils::getSofaPathPrefix()") << "failed to deduce the root path of Sofa from the application path: (" << exePath << ")";
+            // msg_error("Utils::getSofaPathPrefix()") << "failed to deduce the root path of Sofa from the application path: (" << exePath << ")";
```
Any reason to why removing this msg error? Since the line after it returns a wrong sofa prefix path because it is the "Safest thing to return in this case, I guess", we should notify the user of this wrongly path.

Or return an empty string and let the caller handle the error.

jnbrunet replied:
Oh I just saw your previous comment, sorry.

I also got the seg fault while loading sofa in a python environment (without gui) on mac osx. I fear that commenting this line will simply "push" temporary this bug instead of solving it.

jnbrunet replied:
By the way, this bug comes from msg_error called from all those cute sofa static calls when the application (or library) load in memory.

guparan replied:
Did you take some time to investigate the segfault?

guparan replied:
Oh ok. Any idea how to fix it?

jnbrunet replied:
It's on my to do list :)

jnbrunet replied:
https://github.com/sofa-framework/sofa/issues/636 created

guparan committed:
[SofaMacros] ADD transitional macro

sofa_install_get_libraries -> sofa_install_libraries

guparan committed:
[CMake] REMOVE policy settings

As advised by CMake, "projects should be ported to the NEW behavior and not rely on setting a policy to OLD".

guparan committed:
[SofaHelper] Explain commented msg_error in Utils

guparan commented:
Waiting for #642 fix to be complete...

damienmarchal reviewed (commented):
I'm not confident enough with CMakeLists.txt to give a very in-depth review. Maybe it breaks some stuff but as all the build system is a mess this improve things :) 
Let's merge it and see :)

guparan committed:
[CMake] Remove unneeded Qt5::DBus dependency

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_cpack_rules

# Conflicts:
#	SofaKernel/SofaFramework/SofaMacros.cmake

guparan commented:
[ci-build][with-scene-tests]

guparan commented on SofaKernel/SofaFramework/CMakeLists.txt:
```diff
@@ -81,15 +81,18 @@ if(NOT SOFA_NO_OPENGL)
     find_package(GLUT)
-    if(GLUT_FOUND)
+    if(GLUT_FOUND AND SOFAGUI_GLUT)
         if(NOT APPLE)
```
@fredroy You did not remove this block in #598
Should I ?

guparan replied:
Yes.

guparan committed:
[CMake] ADD message about Boost autodection

guparan committed:
[CMake] REMOVE all Qt4 findings

guparan committed:
Remove remaining GLUT references

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_cpack_rules

# Conflicts:
#	SofaGui/CMakeLists.txt
#	applications/sofa/gui/headlessRecorder/HeadlessRecorder.cpp

vmagno committed:
[applications] FIX Install files that were ignored by the `make install` command.

guparan commented:
[ci-build][with-scene-tests]

guparan committed:
Merge remote-tracking branch 'vmagno/fix_install_files' into cmake_cpack_rules

# Conflicts:
#	applications/packages/SofaAllCommonComponents/CMakeLists.txt
#	applications/projects/runSofa/CMakeLists.txt
#	applications/sofa/gui/qt/CMakeLists.txt

guparan committed:
[Geomagic] ADD CMake config file

fredroy reviewed (approved):
If everything compiles... I approve 👍

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_cpack_rules
